Add particle class

gassmoeller opened the pull request:
This is work in progress, but it already allows creating single particles.

davydden commented on include/deal.II/particle/particle.h:
```diff
@@ -0,0 +1,309 @@
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Particle
```
Can u please document this one?

davydden commented on include/deal.II/particle/particle.h:
```diff
@@ -0,0 +1,309 @@
+     * An identifier that denotes the MPI type associated with
+     * types::global_dof_index.
+     */
+#  define ASPECT_PARTICLE_INDEX_MPI_TYPE MPI_UNSIGNED_LONG_LONG
```
Please rename ASPECT to dealii, same below

davydden commented on include/deal.II/particle/particle.h:
```diff
@@ -0,0 +1,309 @@
+
+    /**
+     * Constructor for Particle, creates a particle with the specified
+     * ID at the specified location. Note that Aspect
```
Aspect

davydden commented on include/deal.II/particle/particle.h:
```diff
@@ -0,0 +1,309 @@
+    /**
+     * Constructor for Particle, creates a particle with the specified
+     * ID at the specified location. Note that Aspect
+     * does not check for duplicate particle IDs so the generator must
```
It's not clear what generator in this context, maybe say "user" instead

davydden commented on include/deal.II/particle/particle.h:
```diff
@@ -0,0 +1,309 @@
+     * Set the reference location of this particle. Note that this does not
+     * check whether this is a valid location in the simulation domain.
+     *
+     * @param [in] new_loc The new reference location for this particle.
```
I don't think we use in/out elsewhere in dealii

davydden commented on include/deal.II/particle/particle.h:
```diff
@@ -0,0 +1,309 @@
+   * @ingroup Particle
+   *
+   */
+  template <int dim>
```
Maybe add spacedim as well?

davydden commented on include/deal.II/particle/particle.h:
```diff
@@ -0,0 +1,309 @@
+     * does not check for duplicate particle IDs so the generator must
+     * make sure the IDs are unique over all processes.
+     *
+     * @param[in] new_location Initial location of particle.
```
U don't need "in" here, this is clear from "const"

davydden commented on include/deal.II/particle/particle.h:
```diff
@@ -0,0 +1,309 @@
+    get_location () const;
+
+    /**
+     * Set the reference location of this particle. Note that this does not
```
I guess the note about checks is not needed for dealii

davydden commented on include/deal.II/particle/particle.h:
```diff
@@ -0,0 +1,309 @@
+     * Get the reference location of this particle in its current cell.
+     *
+     * @return The reference location of this particle.
+     */
```
This duplicates the first sentence, maybe just write "Return" above?

davydden replied:
Same below

davydden commented on include/deal.II/particle/particle.h:
```diff
@@ -0,0 +1,309 @@
+
+    /**
+     * Current particle location in the reference cell.
+     * Storing this reduces the number of times we need to compute this
```
This note is probably irrelevant for dealii as well

bangerth commented:
I have a slight preference for `namespace Particles` (plural) for reasons I can not explain. How do others feel? If we changed that, we ought to also change the name of the directory to `particles/`.

drwells commented:
Relates to #4704.

gassmoeller commented:
I have addressed the initial comments. This change is going to happen in small steps so let us know if we do something that conflicts with your intended use @davydden.

Concerning dim vs spacedim. Technically particles themselves are always of type `<0,spacedim>`, and the current `dim` is essentially `spacedim`. Should I rename the template parameter? 
The more interesting case will be the future `ParticleHandler` class, which will likely need to deal with `dim!=spacedim` cases. Currently, I am  not sure what to do in this case. Would the particles need to be projected into the lower dimension? What happens if they move out of the lower dimension?

Concerning the name of the namespace I do not really care, I simply took our current name. However, I can see that `Particles::Particle` is more intuitive than `Particle::Particle`.

gassmoeller commented:
I renamed the namespace and added some tests. Still more tests are necessary, in particular for the property pool.

davydden commented:
You are right about dim-spacedim, u don't need it in the Particle class, I confused myself.

davydden commented on include/deal.II/particles/particle.h:
```diff
@@ -0,0 +1,306 @@
+
+/**
+ * A namespace that contains all classes that are related to the particle
+ * implementation, in particular the fundamental <code>Particle</code> class.
```
If I am not mistaken, you can drop `<code>` here and doxygen will link to `Particle` class automatically. But please check.

bangerth replied:
That's correct.

davydden commented on include/deal.II/particles/particle.h:
```diff
@@ -0,0 +1,306 @@
+   *
+   *
+   * @ingroup Particle
+   *
```
you should add yourself as an author. Please also add a changelog entry.

davydden commented on include/deal.II/particles/particle.h:
```diff
@@ -0,0 +1,306 @@
+     * in the coordinate system of the reference cell.
+     * @param new_id Globally unique ID number of particle.
+     */
+    Particle (const Point<dim> &new_location,
```
i think `new_` in variable names are a bit misleading, i would just call them `location` and `reference_location`.

davydden commented on include/deal.II/particles/particle.h:
```diff
@@ -0,0 +1,306 @@
+     * Set the location of this particle. Note that this does not check
+     * whether this is a valid location in the simulation domain.
+     *
+     * @param [in] new_loc The new location for this particle.
```
drop `in`

bangerth replied:
Why? I think these `[in]`, `[out]`, etc annotations are quite useful.

davydden replied:
see https://github.com/dealii/dealii/pull/4729#discussion_r132124524

davydden commented on include/deal.II/particles/particle.h:
```diff
@@ -0,0 +1,306 @@
+     * afterwards de-serializing the properties by calling the appropriate
+     * constructor Particle(void *&data, const unsigned int data_size);
+     *
+     * @param [in,out] data The memory location to write particle data
```
after doing a search, i realised that the `in/out` question is not really settled. There is already a number of places in documentation which use `in/out`. Sorry for asking you to remove them.
Feel free to use it everywhere you see fit.

davydden commented on include/deal.II/particles/property_pool.h:
```diff
@@ -0,0 +1,88 @@
+  /**
+   * This class manages the memory space in which particles store their
+   * properties. Because this is dynamic memory and every particle needs the
+   * same amount it is more efficient to let this be handled by a central
```
out of curiosity: is it limited to the same amount per particle? What if particles have different types and thus needs to store their properties?

davydden commented on include/deal.II/particles/property_pool.h:
```diff
@@ -0,0 +1,88 @@
+
+    /**
+     * Mark the properties corresponding to the handle @p handle as
+     * deleted. After calling this function calling get_properties() with
```
maybe reword `calling...calling..` to something else?

bangerth commented:
Regarding `dim` and `spacedim` -- I think the class should have both, and store the location as a point `Point<spacedim>` and the reference location as a `Point<dim>`. This is what you would want if you had particles moving around on the surface of a sphere, for example.

bangerth commented:
As everywhere, the default for `spacedim` would of course be `dim`.

luca-heltai commented on include/deal.II/particles/property_pool.templates.h:
```diff
@@ -0,0 +1,84 @@
+{
+  template <typename PropertyType>
+  const typename PropertyPool<PropertyType>::Handle PropertyPool<PropertyType>::invalid_handle = NULL;
+
```
Shouldn't this be a nullptr?

luca-heltai commented on source/particles/particle.cc:
```diff
@@ -0,0 +1,267 @@
+    const types::particle_index *id_data = static_cast<const types::particle_index *> (data);
+    id = *id_data++;
+    const double *pdata = reinterpret_cast<const double *> (id_data);
+
```
Shouldn't these pointers be PropertyType pointers? Maybe I'm missing the point...

luca-heltai replied:
Ah! Now I see.

luca-heltai commented on source/particles/particle.cc:
```diff
@@ -0,0 +1,267 @@
+    // See if there are properties to load
+    const ArrayView<PropertyType> particle_properties = property_pool->get_properties(properties);
+    for (unsigned int i = 0; i < particle_properties.size(); ++i)
+      particle_properties[i] = *pdata++;
```
Is this going to work for arbitrary PropertyType?

luca-heltai commented on source/particles/particle.cc:
```diff
@@ -0,0 +1,267 @@
+      particle_properties[i] = *pdata++;
+
+    data = static_cast<const void *> (pdata);
+  }
```
Maybe using boost::serialize both here and below would simplify life?

tjhei commented on include/deal.II/particles/particle.h:
```diff
@@ -0,0 +1,307 @@
+     * An identifier that denotes the MPI type associated with
+     * types::global_dof_index.
+     */
+#  define PARTICLE_INDEX_MPI_TYPE MPI_UNSIGNED_LONG_LONG
```
why a define and not a constant/constexpr?

tjhei commented on include/deal.II/particles/particle.h:
```diff
@@ -0,0 +1,307 @@
+   *
+   */
+  template <int dim, int spacedim=dim, typename PropertyType = double>
+  class Particle
```
I would like ``Particle`` to just contain a ``T``.

gassmoeller commented:
In a discussion this morning we came to the conclusion that we would like to hide the internal implementation of the particle (in particular the particle properties) inside the particle handler class, and that we therefore need a similar structure to the `TriaIterator` and `TriaAccessor` classes that are already in deal.II. I will start morphing the implementation inside of ASPECT into the new structure and then update this PR when that is finished. This will allow us to better test every change and avoid a lot of `ifdefs` for our existing implementation. This will take some time and I will update this PR afterwards.

davydden commented:
@vishalkenchan please have a look at this PR to make sure you can use it later.

gassmoeller commented:
For reference, the current state of the reconstruction inside ASPECT is tracked here: https://github.com/geodynamics/aspect/issues/1897
The plan would be to morph the functionality, until the classes `Particle`, `ParticleAccessor`, `ParticleIterator`, `ParticleHandler`, and `PropertyPool`can move over into deal.II. The `ParticleHandler` and `ParticleIterator` classes should make sure the internal implementation can be changed later without breaking the interface.

gassmoeller commented:
I have reworked the structure of the Particles namespace to include `ParticleAccessor` and `ParticleIterator` classes. 
The idea behind this is that the `Particle` class is essentially hidden, and if we find that it is more beneficial to change the internal data structures we can do so without breaking the interface. For the same reason I have for now removed the templatized version of the `PropertyPool` and stick to the tested version that only stores `double` properties. The new structure allows to extend this to other types (in any of the ways we discussed before, i.e. using templates, using boost::any, or just storing a block of unidentified data that needs to be interpreted by the user). I think it is simpler to extend this functionality once the basic structure is in place.

I will open another pull request with the `ParticleHandler` class that will illustrate how the classes in this PR are going to be used. Apart from that, I think I addressed all comments and this is ready for another review or merge.

davydden commented on include/deal.II/particles/particle_accessor.h:
```diff
@@ -0,0 +1,210 @@
+     * operation requires a reference to the triangulation.
+     */
+    typename parallel::distributed::Triangulation<dim,spacedim>::cell_iterator
+    get_surrounding_cell (const parallel::distributed::Triangulation<dim,spacedim> &triangulation) const;
```
why does it have to be `p::d::Tria` and not `p::Tria` or even `Tria`?

bangerth replied:
It probably doesn't: `p::d::Tria` is derived from `::Tria`, so if the argument was a reference to the latter, everything would continue to work. Furthermore, the cell iterators of the two classes are the same type, so the function could just return a `::Tria::cell_iterator` without harm.

vishalkenchan commented on include/deal.II/particles/particle_accessor.h:
```diff
@@ -0,0 +1,210 @@
+     * A pointer to the container that stores the particles. Obviously,
+     * this accessor is invalidated if the container changes.
+     */
+    std::multimap<types::LevelInd, Particle<dim,spacedim> >            *map;
```
Do we ever want to allow inheriting from the `Particle` class? Currently the `Particle` class has templated `save()` and `load()` functions which could make it tricky to inherit from. However, if we do want to inherit from `Particle`, we may want this class to be templated with `ParticleType`? as in 
```
template<int dim, int spacedim=dim, typename ParticleType=Particle<dim, spacedim>>
class ParticleAccessor ...
```

davydden replied:
to elaborate on @vishalkenchan comment: it would be good to allow users to extend particles, say in application we work each particle needs to store additionally couple of `Point<dim,spacedim>` and several integer properties. This should work OOB with
```
template<...>
class CustomParticle : public dealii::Particle<dim,spacedim>
{
...
}
```

vishalkenchan commented on include/deal.II/particles/property_pool.h:
```diff
@@ -0,0 +1,103 @@
+     * Returns a new handle that allows accessing the reserved block
+     * of memory.
+     */
+    Handle allocate_properties_array ();
```
Do we want to make this class a bit more abstract? Some of the properties may not be of type `double`, I feel that the properties of particles are hard coded to be a list of `double`s. This looks more like `class ArrayedPropertyPool : public PropertyPool ...`.

Relating this comment to my other comment, may be we want `PropertyPool` class to be more abstract or make `Particle` class inheritable. A more abstract class for particle properties could help users to implement their own versions instead of having to work around this class.

davydden replied:
I suppose a possible cheap work-around is to have additional `IntegerHandler` to store integers. Then such a pool can be used by the main `Particle` class to implement custom particles, but I think it's more cumbersome as compared to polymorphism.

gassmoeller commented:
I addressed the comment about p::d::Tria (and @bangerth is right, there is nothing that should prevent this from being used with shared or serial triangulations, at least it works for p::d::Tria that only uses one process).

Concerning deriving from `Particle` and other types of properties: Let me first discuss why the current structure looks like it does: In our application we do not want to recompile particles whenever their properties change, their properties are usually only known at runtime and there are a lot of possible combinations. That is why we tried to come up with an approach that stores the *basic* particle properties in the class, and everything *optional* or *variable* in additional properties. I admit that the current process is only convenient for everything that can be easily serialized into doubles (like positions, tensors), however in principle the approach works for arbitrary data, i.e. one could cast the stored data to whatever type/structure intended. This can be made more formal in follow-up work, i.e. the type of the handle could be a pointer to a boost::any or a void pointer to a structure. This might be made possible by templatizing `get_properties` so I am not sure deriving from `Particle` is really necessary. That being said: If the only problem are the templatized serialization functions I could remove them without problem for my application. With p::d::Tria I do not use them anyway, but attach their data as additional data to the triangulation (using `register_data_attach` and `notify_ready_to_unpack`). However, I do not see an easy way, how serialization might work in a serial triangulation without these functions. The problem is mostly that particles store the pointer (which is not longer valid after store/load), so they need to store the data themselves, which means they need to know about the type of data. Would you prefer me to remove them for now?

Concerning templatizing `ParticleAccessor` on `ParticleType`: As mentioned above for our purpose that just does not work (there are too many particle types that we can not foresee, and even if we could, we would need to compile tens of different combinations). That being said, I am not opposed to that in general (it might be useful for your case), but as far as I can see that would not break the interface (`Particle<dim,spacedim>` would be the default `ParticleType`). So feel free to extend this functionality once the basics are in.

Would you be ok with the current state and later additions?

davydden commented:
> one could cast the stored data to whatever type/structure intended. This can be made more formal in follow-up work, i.e. the type of the handle could be a pointer to a boost::any or a void pointer to a structure.

I think one can extend it both ways later. For us, a custom particle class is much more convenient to work with.

> If the only problem are the templatized serialization functions I could remove them without problem for my application. With p::d::Tria I do not use them anyway, but attach their data as additional data to the triangulation (using register_data_attach and notify_ready_to_unpack). However, I do not see an easy way, how serialization might work in a serial triangulation without these functions.

Do we know what is the type of archive needed? If there are just a few, one can have
‘’’
public:
virtual void serialize(TypyOne &ar,....)
virtual void serialize(TypeTwo &ar,....)
private:
Template<typename Archive> void
actual_serialization(Archive &ar,...)
‘’’

> As mentioned above for our purpose that just does not work

I guess it does not harm, u just have a single type which is the current Particle. We would also have a single type, just slightly extended.

> So feel free to extend this functionality once the basics are in.

That sounds good to me. I just hope that making this templated won’t break anything for you so that the Particle related classes are flexible enough to cover different needs.

bangerth commented:
I think that rather than threading the data type of the properties through all classes involved here, my preference would be if we could morph the interface in such a way that the `PropertyPool` just returns a block of memory of appropriate size, and the *interpretation* of what that memory corresponds to is left to other parts *outside the immediate particle class and the particle handler*. The point is that at the moment, the particle and particle handler don't actually care what data is being stored. Let's keep it that way.

As @gassmoeller suggested, maybe we can leave this for later patches so that we can make some progress with this patch?

gassmoeller commented:
> Do we know what is the type of archive needed? If there are just a few, one can have
‘’’
public:
virtual void serialize(TypyOne &ar,....)
virtual void serialize(TypeTwo &ar,....)
private:
Template void
actual_serialization(Archive &ar,...)
‘’’

As far as I know it would need to be instantiated for all of the `boost::archive` types, but I have to admit I never quite read through all of those mechanisms myself. These are the types that are listed in boost (http://www.boost.org/doc/libs/1_65_1/libs/serialization/doc/index.html):

```
boost::archive::text_oarchive // saving
boost::archive::text_iarchive // loading

// a portable text archive using a wide character stream
boost::archive::text_woarchive // saving
boost::archive::text_wiarchive // loading

// a portable XML archive
boost::archive::xml_oarchive // saving
boost::archive::xml_iarchive // loading

// a portable XML archive which uses wide characters - use for utf-8 output
boost::archive::xml_woarchive // saving
boost::archive::xml_wiarchive // loading

// a non-portable native binary archive
boost::archive::binary_oarchive // saving
boost::archive::binary_iarchive // loading
```

But could you explain to me why a derived class with additional members would have trouble using these functions? According to http://www.boost.org/doc/libs/1_65_1/libs/serialization/doc/index.html the derived class should be able to call the serialization functions of the base class (not quite, but effectively so), and if your derived class does not use properties at all it should also have no problems if `properties` is a pointer to double, because no properties are serialized in the base class if there are no properties.


> That sounds good to me. I just hope that making this templated won’t break anything for you so that the Particle related classes are flexible enough to cover different needs.

I hope that they are flexible enough by now. I will try to include testcases into #5125 that are close to my application case so that we notice if something breaks. If something still breaks we will notice soon enough, since this code is already used by a number of our ASPECT users and some of them regularly work with deal.II dev.


@luca-heltai you requested changes to the original version before the rework. Do you agree with proceeding with this version and improve it later on (like making the `reference_location` another optional property instead of base property)?

davydden commented:
> As @gassmoeller suggested, maybe we can leave this for later patches so that we can make some progress with this patch?

I am perfectly fine with that.

> But could you explain to me why a derived class with additional members would have trouble using these functions?

Initially I was thinking about the case where some function takes a base class Particle and the derived class is provided instead. But if all the relevant parts are templated with a ParticleType, then we should not have any problems with serialization as long as user class implements the required interface.

> I will try to include testcases into #5125 that are close to my application case so that we notice if something breaks

That sounds good. Then we shall be sure that whatever improvements are to come, it won’t break the original usage case.

davydden commented:
@luca-heltai ping.

bangerth commented:
Is this superseded by #5125 ?

davydden commented:
> Is this superseded by #5125 ?

I don’t think so, 5125 is based on this and adds extra functionality.

I think we should just merge this one to get going

gassmoeller commented:
No this is ready in its current form. #5125 adds a new class `ParticleHandler` and is a follow-up. I talked to @luca-heltai two weeks ago about the approach and he agreed, so I suppose we do not need to delay this PR further.

davydden commented:
/run-tests

davydden commented:
@luca-heltai i think we are not merging this PR only because you requested changes, can you please give it a look and approve if it's OK?

luca-heltai commented:
Yes. I'm fine with this as is. Go ahead and merge. I cannot, because there are (unrelated) failing tests
